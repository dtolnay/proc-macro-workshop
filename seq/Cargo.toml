[package]
name = "seq"
version = "0.0.0"
autotests = false
edition = "2021"
publish = false

[lib]
proc-macro = true

[[test]]
name = "tests"
path = "tests/progress.rs"

[dev-dependencies]
trybuild = { version = "1.0.49", features = ["diff"] }

[dependencies]
proc-macro2 = "1.0"
quote = "1.0"
syn = "2.0"
