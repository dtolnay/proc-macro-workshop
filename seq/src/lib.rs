mod bounds;
mod expand;
mod parse;

use crate::parse::Input;
use proc_macro::TokenStream;
use syn::parse_macro_input;

#[proc_macro]
pub fn seq(input: TokenStream) -> TokenStream {
    let input = parse_macro_input!(input as Input);
    let expanded = expand::expand(input).unwrap_or_else(|err| err.to_compile_error());
    TokenStream::from(expanded)
}
