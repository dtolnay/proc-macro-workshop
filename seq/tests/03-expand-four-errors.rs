// Now construct the generated code! Produce the output TokenStream by repeating
// the loop body the correct number of times as specified by the loop bounds and
// replacing the specified identifier with the loop counter.
//
// The invocation below will need to expand to a TokenStream containing:
//
//     compile_error!(concat!("error number ", stringify!(0)));
//     compile_error!(concat!("error number ", stringify!(1)));
//     compile_error!(concat!("error number ", stringify!(2)));
//     compile_error!(concat!("error number ", stringify!(3)));
//
// This test is written as a compile_fail test because our macro isn't yet
// powerful enough to do anything useful. For example if we made it generate
// something like a function, every one of those functions would have the same
// name and the program would not compile.
//
// Resources:
//
//   - Struct to represent integer literals
//     https://docs.rs/syn/latest/syn/struct.LitInt.html
//
//   - Creating new token streams from a sequence of token trees
//     https://docs.rs/proc-macro2/latest/proc_macro2/struct.TokenStream.html#impl-FromIterator%3CTokenTree%3E-for-TokenStream

use seq::seq;

seq!(N in 0..4 {
    compile_error!(concat!("error number ", stringify!(N)));
});

fn main() {}
