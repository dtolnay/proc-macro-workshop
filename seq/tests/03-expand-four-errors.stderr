error: error number 0
   |
20 |     compile_error!(concat!("error number ", stringify!(N)));
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: error number 1
   |
20 |     compile_error!(concat!("error number ", stringify!(N)));
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: error number 2
   |
20 |     compile_error!(concat!("error number ", stringify!(N)));
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: error number 3
   |
20 |     compile_error!(concat!("error number ", stringify!(N)));
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to 4 previous errors
